### DISCUSSION TITLE
Poor frame pacing due to inconsistent sleep

### DISCUSSION THREAD

**Discussion by alion02** [ppy/osu#19826] [Q&A]
### Type

Performance

### Bug description

With the [removal of truly unlimited framerate](https://github.com/ppy/osu-framework/pull/5235), the game no longer offers me a sufficiently smooth, stutter-free experience. Upon further investiagtion I discovered that the underlying reason for the stutters are inconsistencies in sleep times caused by [insufficient OS timer resolution](https://stackoverflow.com/a/22862989).

Assuming 1 ms resolution, each sleep call will typically sleep for anywhere between 1 and 2 milliseconds. Worse still, sleeping doesn't necessarily align well between different threads when running osu! in multithreaded mode.

This means that the combined sleep time for the update and draw threads may reach around 4 ms, which is enough to straddle (i.e. skip) up to 2 frames on [some monitors](https://www.nvidia.com/en-us/geforce/technologies/360-hz/). Note that you don't need such an extreme display to perceive the stutter. While I doubt it's noticeable on a plain 60hz screen, a relatively commonplace nowadays 120hz one, on which the speed of an object may fluctuate by up to around 50%, is plenty. The same logic applies to the input and update threads, which may affect high-accuracy content.

[There](https://www.pinvoke.net/default.aspx/ntdll/NtSetTimerResolution.html) [are](http://undocumented.ntinternals.net/index.html?page=UserMode%2FUndocumented%20Functions%2FNT%20Objects%2FThread%2FNtDelayExecution.html) [ways](https://docs.microsoft.com/en-us/dotnet/api/system.threading.thread.spinwait) to reduce the severity of or resolve this issue (at least on Windows, I don't have experience with other OSes) that don't involve bringing back truly unlimited framerate (and would also benefit non-draw threads).

### Screenshots or videos

![Frametime graph](https://user-images.githubusercontent.com/22531550/185270779-e230cdfe-f616-438d-8bd1-79197db57d51.png)

### Version

2022.816.0-lazer

### Logs

[runtime.log](https://github.com/ppy/osu/files/9366697/runtime.log)
[network.log](https://github.com/ppy/osu/files/9366698/network.log)
[performance.log](https://github.com/ppy/osu/files/9366699/performance.log)
[updater.log](https://github.com/ppy/osu/files/9366700/updater.log)

**Comment by peppy** [comment]
Firstly, you've made a lot of assumption, which I while conceptually are valid, I don't believe are correct in lazer's current implementation. I'm aware of the issues that can arise, but in extensive testing am 100% sure that at least on my isolated setups, there is no such issue. We already have very good frame alignment due to the mutex blocking between update and draw thread.

So basically, I'd prefer you report what you are seeing rather than provide technical details / explanations that are likely not the full story, or may be a red herring altogether.

- Are you noticing this visually? If so, please provide a slow motion recording (120-240fps) using your phone camera (or a screen capture if that is enough to convey the issue) showing it in action.
- Alternatively, can you confirm using the built-in latency certifier that you can go above 1000hz? That would be what you are implying here.

**Reply by alion02** [reply]
- I am noticing this visually, I even created a separate branch to continue playing on soon after the unlimited framerate removal.
- I have tried to get either a slo-mo phone recording or a screen recording but both sort of failed.
- We had a brief e-mail exchange with video proof that I certified at 1000hz.

**Reply by peppy** [reply]
I'll wait to see a recording of this to further understand the issue. Take your time.

If it's visual you should also be able to just capture using OBS, for what it's worth.

**Reply by alion02** [reply]
Note that, due to the sleep lasting anywhere between 1 and 2 milliseconds, the practical highest consistent framerate that can possibly be achieved here is 500 (without making any assumptions about osu!, just that the frametime graph is accurate).

**Reply by alion02** [reply]
Also, I'm fairly confident in my diagnosis for one more reason - I've created a separate program to set the timer resolution to 0.5 ms, which noticeably reduced the stutter.

I'll do my best to capture the problem, but OBS severely impacts performance while my phone seemingly inserts stutters into the recording.

**Reply by alion02** [reply]
Some progress - compare the original frametime graph screenshot with this one, which was captured during a period of less noticeable stuttering:
![less stutter](https://user-images.githubusercontent.com/22531550/185282653-02972dcf-ca95-480f-a0c9-5990338d8d86.png)
In the original, the update and draw frametime graphs approximately have a period of 3, with longer sleeps, while here it fell into a period of 2 with shorter sleeps.

osu! seems to semi-randomly switch between these two frametime graphs. I haven't been able to observe the period 3 graph with the timer resolution forced to 0.5 ms by my external application.

**Reply by peppy** [reply]
Thanks for the updates, but it will get very noisy if you keep posting in new posts. Please edit your previous posts with new information and only make a follow-up post if providing the original requested information (video showing the issue). Also when you do so, please include your PC and monitor specs in the same post.

**Reply by alion02** [reply]
Specs:
- Ryzen 3700X
- GTX 1080
- SSD
- 1080p360hz
- Win10 21H1

[The video.](https://anonymfile.com/Dj0d/2022-08-18-20-27-54.mp4) Running osu! at 720p because neither my GPU nor CPU liked me trying to encode a 1080p360 video in real time, and scaling it in OBS would've resulted in a blurry frametime graph.

The graph looks more chaotic than in either of the screenshots because of OBS, and the experience is slightly worse than anything that can occur without it running. Looking at the video frame by frame, it's clear that the objects don't have a consistent speed. This is especially noticeable on high velocity objects. That said, the stutter is much less egregious than theorized - there are hardly any dropped frames, let alone two consecutive - must be the update-draw thread synchronization you mentioned.

One rather important thing I didn't realize at first is that this issue (to my knowledge) only practically affects osu!taiko. I tried osu!std and osu!mania and couldn't perceive any stuttering in either of them. I don't know about other players, but I precisely follow objects with my eyes in taiko, which means that, compared to std approach circles or slider balls or mania notes, it's much easier to notice any inconsistencies in their movement.

**Reply by peppy** [reply]
Thanks, I'll analyse this over the weekend, but what you are saying now more closely matches with what I saw in my original testing, and one of the remaining points on my list to fix (there's already a path forward to improve this but I'm waiting for our renderer pipeline to stabilise before considering it). As I mentioned, we are getting *rough* frame pacing by aligning the draw and update thread, but this can further be improved with driver hooks and more precise update scheduling.

Also, the reason you're noticing more in taiko is likely due to screen tearing (or put another way, the direction of updates on an lcd) occurring on the same axis as movement, if anything. Depending on the screen technology (OLED, IPS, etc.) this can get better or worse (I've confirmed this in my own testing across multiple devices, with some devices managing to make 60hz feel as good as perfectly smooth 120hz just due to the refresh method).

To confirm, you're running with all vsync modes off and on an LCD display? And your display is full 360hz?

**Reply by alion02** [reply]
Yes, native 360hz LCD without any v-sync. As for tearing, OBS shouldn't be capturing any, and a video player shouldn't be adding any, and the recording doesn't seem any smoother to me than the osu! client.

**Reply by alion02** [reply]
Any updates on this? And is there anything else I can do to help?

**Reply by peppy** [reply]
turning on vsync should help you if you're looking for smoothness.

**Reply by alion02** [reply]
There's a separate issue with that... I get locked 180 fps most of the time with v-sync on. And no, that's not my system to blame - I used to get >1000 before the forced cap, and it's a problem unique to osu!lazer, since stable runs just fine with v-sync.

**Reply by peppy** [reply]
Again, that's the only thing I can offer. No one else is seeing these kinds of issues so it needs more investigation into your system specifically, but I would suggest waiting until direct x support comes online rather than posting further in this thread.
